### PR TITLE
NodeId implementation

### PR DESCRIPTION
**Pull request by evanxg852000**
Implements the concept of NodeId described [here](https://www.notion.so/Cluster-Node-Identification-83c2ba2568bc4a199aaf73d04bb98066) on the ScuttleButt .

### CODE REVIEW

File: scuttlebutt/src/lib.rs [hunk 1]
```diff
@@ -39,15 +39,53 @@ use tracing::{debug, error};
...
 
 pub type Version = u64;
 
-pub type NodeId = String;
+// pub type NodeId = String;
+
+#[derive(Default, Clone, Hash, Eq, PartialEq, PartialOrd, Ord, Debug, Serialize, Deserialize)]
+pub struct NodeId {
+    id: String,
+    gossip_public_address: String,
```
Comment (fulmicoton) [inline scuttlebutt/src/lib.rs:55]: why not SocketAddr ?

File: scuttlebutt/src/lib.rs [hunk 2]
```diff
@@ -39,15 +39,53 @@ use tracing::{debug, error};
...
 
 /// Map key for the heartbeat node value.
 pub(crate) const HEARTBEAT_KEY: &str = "heartbeat";
 
 pub type Version = u64;
 
-pub type NodeId = String;
+// pub type NodeId = String;
+
+#[derive(Default, Clone, Hash, Eq, PartialEq, PartialOrd, Ord, Debug, Serialize, Deserialize)]
```
Comment (fulmicoton) [inline scuttlebutt/src/lib.rs:52]: We need doc here. Copy pasting your doc is ok.

File: scuttlebutt/src/lib.rs [hunk 3]
```diff
@@ -39,15 +39,53 @@ use tracing::{debug, error};
...
+pub struct NodeId {
+    id: String,
+    gossip_public_address: String,
+}
+
+impl NodeId {
+    fn new(id: &str, gossip_public_address: &str) -> Self {
```
Comment (fulmicoton) [inline scuttlebutt/src/lib.rs:59]: That should be `String`, or T: ToString if it is too painful in tests.

File: scuttlebutt/src/lib.rs (hunk 1 above)
Comment (fulmicoton) [inline scuttlebutt/src/lib.rs:55]: why did you remove `node_generation`

Comment (evanxg852000) [reply scuttlebutt/src/lib.rs:55]: hmm, I needed to implement Default, not sure of a good default SocketAddr and this is not an option really.  Also, this made testing easier because I could `NodeId::from("node-one")` without worrying of passing a valid SocketAdrr.

File: scuttlebutt/src/lib.rs (hunk 2 above)
Comment (evanxg852000) [reply scuttlebutt/src/lib.rs:52]: Adding doc, also part of the doc will be in Quickwit where QuickwitNodeId is defined.

File: scuttlebutt/src/lib.rs (hunk 1 above)
Comment (evanxg852000) [reply scuttlebutt/src/lib.rs:87]: It will be available from the client-side (Quickwit). See the docs
while writing the docs, I noticed we can just use `{node_unique_id}/{node_generation}` instead of `{node_unique_id}/{gossip_public_address}/{generation}`
https://github.com/quickwit-oss/scuttlebutt/pull/22/files#diff-9a1c8c0b2fe22d61843aec8d062321324f330717737a4880c634fc3c5a7b43fbR74

**Review by fulmicoton — APPROVED** [review]
